src/log-store/src/raft_engine/backend.rs

```diff
@@ -0,0 +1,548 @@
+                    .map(|value| KeyValue {
+                        key: key.clone(),
+                        value,
+                    })
+                    .collect();
+                Ok(TxnOpResponse::ResponseGet(RangeResponse {
+                    kvs,
+                    more: false,
+                }))
+            }
+
+            TxnOp::Delete(key) => {
+                let prev_value = engine_get(&engine, &key)?;
+                batch.delete(SYSTEM_NAMESPACE, key);
+
+                let prev_kvs: Vec<_> = prev_value.into_iter().collect();
+                let deleted = prev_kvs.len() as i64;
+
+                Ok(TxnOpResponse::ResponseDelete(DeleteRangeResponse {
+                    deleted,
+                    prev_kvs,
+                }))
+            }
+        };
+
+        let responses = if succeeded { success } else { failure }
+            .into_iter()
+            .map(do_txn)
+            .collect::<error::Result<_>>()?;
+
+        engine.write(&mut batch, false).context(RaftEngineSnafu)?;
+
+        Ok(TxnResponse {
+            succeeded,
+            responses,
+        })
+    }
+}
+
+#[async_trait::async_trait]
+impl KvBackend for RaftEngineBackend {
+    fn name(&self) -> &str {
+        "RaftEngineBackend"
+    }
+
+    fn as_any(&self) -> &dyn Any {
+        self
+    }
+
+    async fn range(&self, req: RangeRequest) -> Result<RangeResponse, Self::Error> {
+        let mut res = vec![];
+        self.engine
+            .read()
+            .unwrap()
+            .scan_raw_messages(
+                SYSTEM_NAMESPACE,
+                Some(&req.key),
+                Some(&req.range_end),
+                false,
+                |key, value| {
+                    res.push(KeyValue {
+                        key: key.to_vec(),
+                        value: value.to_vec(),
+                    });
+                    true
+                },
+            )
+            .context(RaftEngineSnafu)?;
+        Ok(RangeResponse {
+            kvs: res,
+            more: false,
+        })
+    }
+
+    async fn put(&self, req: PutRequest) -> Result<PutResponse, Self::Error> {
+        let PutRequest {
+            key,
+            value,
+            prev_kv,
+        } = req;
+
+        let mut prev = None;
+        let engine = self.engine.read().unwrap();
+        if prev_kv {
+            prev = engine_get(&engine, &key)?;
+        }
+        engine_put(&engine, key, value)?;
+        Ok(PutResponse { prev_kv: prev })
+    }
+
+    async fn batch_put(&self, req: BatchPutRequest) -> Result<BatchPutResponse, Self::Error> {
+        let BatchPutRequest { kvs, prev_kv } = req;
+        let mut batch = LogBatch::with_capacity(kvs.len());
+
+        let mut prev_kvs = if prev_kv {
+            Vec::with_capacity(kvs.len())
+        } else {
+            vec![]
+        };
+
+        for kv in kvs {
+            if prev_kv && let Some(kv) = engine_get(&self.engine.read().unwrap(), &kv.key)? {
```

Comment:
Is it better to lift the acquisition of read lock out of the "for" loop?

Reply:
> Is it better to lift the acquisition of read lock out of the "for" loop?

Acquiring lock per key gives other operations that require write lock a chance to execute. But I think it would not make a noticeable difference. I will change it.

Reply:
fixed in [257c7a6](https://github.com/GreptimeTeam/greptimedb/pull/2280/commits/257c7a6d3cd42ac501dfecfa3533502c5011f0e9#diff-74b285f40ce368ca6fe6df02183ca1e98c68112554e907aa9e7271ee6ca2d6feR175-R177)